// @vendure/core/dist/config/entity/entity-duplicators/product-duplicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.productDuplicator = void 0;
const generated_types_1 = require("@vendure/common/lib/generated-types");
const typeorm_1 = require("typeorm");
const common_1 = require("../../../common");
const errors_1 = require("../../../common/error/errors");
const transactional_connection_1 = require("../../../connection/transactional-connection");
const product_entity_1 = require("../../../entity/product/product.entity");
const product_option_group_entity_1 = require("../../../entity/product-option-group/product-option-group.entity");
const product_variant_entity_1 = require("../../../entity/product-variant/product-variant.entity");
const product_option_group_service_1 = require("../../../service/services/product-option-group.service");
const product_option_service_1 = require("../../../service/services/product-option.service");
const product_variant_service_1 = require("../../../service/services/product-variant.service");
const product_service_1 = require("../../../service/services/product.service");
const entity_duplicator_1 = require("../entity-duplicator");
let connection;
let productService;
let productVariantService;
let productOptionGroupService;
let productOptionService;
/**
 * @description
 * Duplicates a Product and its associated ProductVariants.
 */
exports.productDuplicator = new entity_duplicator_1.EntityDuplicator({
    code: 'product-duplicator',
    description: [
        {
            languageCode: generated_types_1.LanguageCode.en,
            value: 'Default duplicator for Products',
        },
    ],
    requiresPermission: [generated_types_1.Permission.CreateProduct, generated_types_1.Permission.CreateCatalog],
    forEntities: ['Product'],
    args: {
        includeVariants: {
            type: 'boolean',
            defaultValue: true,
            label: [{ languageCode: generated_types_1.LanguageCode.en, value: 'Include variants' }],
        },
    },
    init(injector) {
        connection = injector.get(transactional_connection_1.TransactionalConnection);
        productService = injector.get(product_service_1.ProductService);
        productVariantService = injector.get(product_variant_service_1.ProductVariantService);
        productOptionGroupService = injector.get(product_option_group_service_1.ProductOptionGroupService);
        productOptionService = injector.get(product_option_service_1.ProductOptionService);
    },
    async duplicate({ ctx, id, args }) {
        var _a;
        const product = await connection.getEntityOrThrow(ctx, product_entity_1.Product, id, {
            relations: {
                featuredAsset: true,
                assets: true,
                channels: true,
                facetValues: {
                    facet: true,
                },
                optionGroups: {
                    options: true,
                },
            },
        });
        const translations = product.translations.map(translation => {
            return {
                name: translation.name + ' (copy)',
                slug: translation.slug + '-copy',
                description: translation.description,
                languageCode: translation.languageCode,
                customFields: translation.customFields,
            };
        });
        const productInput = {
            featuredAssetId: (_a = product.featuredAsset) === null || _a === void 0 ? void 0 : _a.id,
            enabled: false,
            assetIds: product.assets.map(value => value.assetId),
            facetValueIds: product.facetValues.map(value => value.id),
            translations,
            customFields: product.customFields,
        };
        const duplicatedProduct = await productService.create(ctx, productInput);
        if (args.includeVariants) {
            const productVariants = await connection.getRepository(ctx, product_variant_entity_1.ProductVariant).find({
                where: {
                    productId: id,
                    deletedAt: (0, typeorm_1.IsNull)(),
                },
                relations: {
                    options: {
                        group: true,
                    },
                    assets: true,
                    featuredAsset: true,
                    stockLevels: true,
                    facetValues: true,
                    productVariantPrices: true,
                    taxCategory: true,
                },
            });
            if (product.optionGroups && product.optionGroups.length) {
                for (const optionGroup of product.optionGroups) {
                    const newOptionGroup = await productOptionGroupService.create(ctx, {
                        code: optionGroup.code,
                        translations: optionGroup.translations.map(translation => {
                            return {
                                languageCode: translation.languageCode,
                                name: translation.name,
                                customFields: translation.customFields,
                            };
                        }),
                    });
                    const options = optionGroup.options.map(option => {
                        return {
                            code: option.code,
                            productOptionGroupId: newOptionGroup.id,
                            translations: option.translations.map(translation => {
                                return {
                                    languageCode: translation.languageCode,
                                    name: translation.name,
                                    customFields: translation.customFields,
                                };
                            }),
                        };
                    });
                    if (options && options.length) {
                        for (const option of options) {
                            const newOption = await productOptionService.create(ctx, newOptionGroup, option);
                            newOptionGroup.options.push(newOption);
                        }
                    }
                    await productService.addOptionGroupToProduct(ctx, duplicatedProduct.id, newOptionGroup.id);
                }
            }
            const newOptionGroups = await connection.getRepository(ctx, product_option_group_entity_1.ProductOptionGroup).find({
                where: {
                    product: { id: duplicatedProduct.id },
                },
                relations: {
                    options: true,
                },
            });
            const variantInput = productVariants.map((variant, i) => {
                var _a, _b, _c, _d, _e;
                const options = variant.options.map(existingOption => {
                    var _a;
                    const newOption = (_a = newOptionGroups
                        .find(og => og.code === existingOption.group.code)) === null || _a === void 0 ? void 0 : _a.options.find(o => o.code === existingOption.code);
                    if (!newOption) {
                        throw new errors_1.InternalServerError(`An error occurred when creating option ${existingOption.code}`);
                    }
                    return newOption;
                });
                const price = (_b = (_a = variant.productVariantPrices.find(p => (0, common_1.idsAreEqual)(p.channelId, ctx.channelId))) === null || _a === void 0 ? void 0 : _a.price) !== null && _b !== void 0 ? _b : (_c = variant.productVariantPrices[0]) === null || _c === void 0 ? void 0 : _c.price;
                return {
                    productId: duplicatedProduct.id,
                    price: price !== null && price !== void 0 ? price : variant.price,
                    sku: `${variant.sku}-copy`,
                    stockOnHand: 1,
                    featuredAssetId: (_d = variant.featuredAsset) === null || _d === void 0 ? void 0 : _d.id,
                    taxCategoryId: (_e = variant.taxCategory) === null || _e === void 0 ? void 0 : _e.id,
                    useGlobalOutOfStockThreshold: variant.useGlobalOutOfStockThreshold,
                    trackInventory: variant.trackInventory,
                    assetIds: variant.assets.map(value => value.assetId),
                    translations: variant.translations.map(translation => {
                        return {
                            languageCode: translation.languageCode,
                            name: translation.name,
                        };
                    }),
                    optionIds: options.map(option => option.id),
                    facetValueIds: variant.facetValues.map(value => value.id),
                    stockLevels: variant.stockLevels.map(stockLevel => ({
                        stockLocationId: stockLevel.stockLocationId,
                        stockOnHand: stockLevel.stockOnHand,
                    })),
                };
            });
            const duplicatedProductVariants = await productVariantService.create(ctx, variantInput);
            duplicatedProduct.variants = duplicatedProductVariants;
        }
        return duplicatedProduct;
    },
});
